### PR TITLE
extend particle handle

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- add method `setHandleInvalid()`
- add method `isHandleValid()`
- set internal point to non `const` to allow the state change to invalid

This is a pre work to support the new filter functors. The new functor concepts are needed to use the PIConGPU manipulator together with the lock step programming model and alpaka.

### CODE REVIEW

File: src/libPMacc/include/particles/memory/dataTypes/Particle.hpp [hunk 1]
```diff
@@ -81,7 +81,31 @@ struct Particle : public InheritLinearly<typename T_FrameType::MethodsList>
      * ATTENTION: The pointer must be the last member to avoid local memory usage
      *            https://github.com/ComputationalRadiationPhysics/picongpu/pull/762
      */
-    PMACC_ALIGN(frame, FrameType* const);
+    PMACC_ALIGN(frame, FrameType*);
+
+    /** set particle handle to invalid
+     *
+     * This method set the particle handle to invalid. It is possible to proof with
```
Comment (ax3l) [inline src/libPMacc/include/particles/memory/dataTypes/Particle.hpp]: set**s** the particle handle...

File: src/libPMacc/include/particles/memory/dataTypes/Particle.hpp [hunk 2]
```diff
@@ -81,7 +81,31 @@ struct Particle : public InheritLinearly<typename T_FrameType::MethodsList>
...
+     */
+    HDINLINE void setHandleInvalid()
+    {
+        frame = nullptr;
+    }
+
+    /** check if particle handle is valid
+     *
+     * A valid particle handle means that the memory behind the handle can be used
+     * save, it not means that the particle multiMask is `>= 1`.
```
Comment (ax3l) [inline src/libPMacc/include/particles/memory/dataTypes/Particle.hpp]: - savely

- [...]. An invalid handle means that particle exists but its multiMask is `>= 1`, which means ...

File: src/libPMacc/include/particles/memory/dataTypes/Particle.hpp (hunk 1 above)
Comment (ax3l) [inline src/libPMacc/include/particles/memory/dataTypes/Particle.hpp]: proof -> test

File: src/libPMacc/include/particles/memory/dataTypes/Particle.hpp [hunk 3]
```diff
@@ -81,7 +81,31 @@ struct Particle : public InheritLinearly<typename T_FrameType::MethodsList>
...
      */
-    PMACC_ALIGN(frame, FrameType* const);
+    PMACC_ALIGN(frame, FrameType*);
+
+    /** set particle handle to invalid
+     *
+     * This method set the particle handle to invalid. It is possible to proof with
+     * the method isHandleValid if the particle is valid.
+     * If the particle is set to invalid it is not allowed to call any method other
+     * than isHandleValid or setHandleInvalid.
```
Comment (ax3l) [inline src/libPMacc/include/particles/memory/dataTypes/Particle.hpp]: , but it does not mean the particle is deactivated outside of ...

File: src/libPMacc/include/particles/memory/dataTypes/Particle.hpp (hunk 2 above)
Comment (ax3l) [inline src/libPMacc/include/particles/memory/dataTypes/Particle.hpp]: A valid handle does not mean that the particle's multiMask is valid (>=1).

**Comment by psychocoderHPC** [comment]
updated

**Pull request merged**